synapse/cores/common.py

```diff
@@ -1938,3 +1912,68 @@ def _tufosByInetCidr(self, prop, valu, limit=None):
 
         return self.getTufosBy('range', prop, (ipv4addr, ipv4addr+mask), limit=limit)
 
+    #############################################################
+    # support typelib persistence
+    #############################################################
+
+    # overrides: synapse.lib.types.TypeLib.addType
+    def addType(self, typ):
+        s_types.TypeLib.addType(self, typ)
+        self.formTufoByFrob('syn:type', typ.name)
+
+    # overrides: synapse.lib.types.TypeLib.addSubType
+    def addSubType(self, name, subof, **info):
+        s_types.TypeLib.addSubType(self, name, subof, **info)
+        base = self.getTufoByFrob('syn:type', subof)
+
+        subinfo = {}
+        for propname, propval in base[1].items():
+            if not propname.startswith('syn:type:'):
+                continue
+            propname = propname.partition('syn:type')[2]
+            subinfo[propname] = propval
+
+        for propname, propval in info.items():
+            subinfo[propname] = propval
+
+        subinfo['base'] = subof
+
+        if 'fields' in subinfo:
+            fields = subinfo.pop('fields')
+            subinfo['fields'] = len(fields)
+            for i, (fieldname, fieldtype) in enumerate(fields):
+                k = 'field:%d:%s' % (i, fieldname)
+                subinfo[k] = fieldtype
```

Comment:
what this means is that:

    core.addSubType('dns:a','comp',fields=(('fqdn','inet:fqdn'),('ipv4','inet:ipv4'),('time','time:epoch')))

is represented:

    syn:type="dns:a"
    syn:type:base="comp"
    syn:type:fields=3
    syn:type:field:0:fqdn="inet:fqdn"
    syn:type:field:1:ipv4="inet:ipv4"
    syn:type:field:2:time="time:epoch"